compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateAIEWorkgroup.cpp

```diff
@@ -70,24 +70,58 @@ LogicalResult workgroupBuildForDmaCpyNdOp(
     Block::iterator controlCodeBegin, Block::iterator controlCodeEnd) {
   LLVM_DEBUG(llvm::dbgs() << "workgroupBuild [amdaie.dma_cpy_nd] Start\n");
   Attribute sourceMemSpace = dmaOp.getSourceObjectFifo().getMemorySpace();
+  Attribute targetMemSpace = dmaOp.getTargetObjectFifo().getMemorySpace();
   Location loc = rewriter.getUnknownLoc();
   SmallVector<OpFoldResult> empty;
+
+  SmallVector<OpFoldResult> circularDmaTargetOffsets = empty;
+  SmallVector<OpFoldResult> circularDmaTargetSizes = empty;
+  SmallVector<OpFoldResult> circularDmaTargetStrides = empty;
+  SmallVector<OpFoldResult> circularDmaSourceOffsets = empty;
+  SmallVector<OpFoldResult> circularDmaSourceSizes = empty;
+  SmallVector<OpFoldResult> circularDmaSourceStrides = empty;
+
+  SmallVector<OpFoldResult> ipuDmaTargetOffsets = dmaOp.getTargetMixedOffsets();
+  SmallVector<OpFoldResult> ipuDmaTargetSizes = dmaOp.getTargetMixedSizes();
+  SmallVector<OpFoldResult> ipuDmaTargetStrides = dmaOp.getTargetMixedStrides();
+  SmallVector<OpFoldResult> ipuDmaSourceOffsets = dmaOp.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> ipuDmaSourceSizes = dmaOp.getSourceMixedSizes();
+  SmallVector<OpFoldResult> ipuDmaSourceStrides = dmaOp.getSourceMixedStrides();
+  if (!sourceMemSpace) {
+    // Check if the source of DmaCpyNd op is from L3 - then addressing will be
+    // controlled by the uController.
+    circularDmaTargetOffsets = ipuDmaTargetOffsets;
+    circularDmaTargetSizes = ipuDmaTargetSizes;
+    circularDmaTargetStrides = ipuDmaTargetStrides;
+
+    ipuDmaTargetOffsets = empty;
+    ipuDmaTargetSizes = empty;
+    ipuDmaTargetStrides = empty;
+  } else if (!targetMemSpace) {
+    // Check if the target of DmaCpyNd op is from L3 - then addressing will be
+    // controlled by the circular DMA at shim tile.
+    circularDmaSourceOffsets = ipuDmaSourceOffsets;
+    circularDmaSourceSizes = ipuDmaSourceSizes;
+    circularDmaSourceStrides = ipuDmaSourceStrides;
+
+    ipuDmaSourceOffsets = empty;
+    ipuDmaSourceSizes = empty;
+    ipuDmaSourceStrides = empty;
+  }
+  // In case of transfers between L1/L2 - the addressing is always controlled by
+  // the uController.
```

Comment:
Incorrect, L1/L2 transfers are always part of the AIE configuration for now. However, these should not go through this function as these are all `circular_dma_cpy_nd` at the moment and I would just add a check at the top to error out in that case until we want to handle it. See comment above. This probably means adjusting the tests below as there is a check with `dma_cpy_nd` between L1/L2.